dali/pipeline/operators/resize/resize.cu

```diff
@@ -12,155 +12,105 @@
+  SetupResamplingParams();
 }
 
 template<>
 void Resize<GPUBackend>::SetupSharedSampleParams(DeviceWorkspace* ws) {
   auto &input = ws->Input<GPUBackend>(0);
+
   DALI_ENFORCE(IsType<uint8>(input.type()), "Expected input data as uint8.");
+  if (input.GetLayout() != DALI_UNKNOWN) {
+    DALI_ENFORCE(input.GetLayout() == DALI_NHWC,
+                 "Resize expects interleaved channel layout (NHWC)");
+  }
 
   for (int i = 0; i < batch_size_; ++i) {
     vector<Index> input_shape = input.tensor_shape(i);
     DALI_ENFORCE(input_shape.size() == 3, "Expects 3-dimensional image input.");
 
     per_sample_meta_[i] = GetTransformMeta(spec_, input_shape, ws, i, ResizeInfoNeeded());
+    resample_params_[i] = GetResamplingParams(per_sample_meta_[i]);
+  }
+  SetupResamplingParams();
+}
+
+template <int ndim>
+void ToDimsVec(std::vector<Dims> &dims_vec, const kernels::TensorListShape<ndim> &tls) {
+  const int dim = tls.sample_dim();
+  const int N = tls.num_samples();
+  dims_vec.resize(N);
+
+  for (int i = 0; i < N; i++) {
+    dims_vec[i].resize(dim);
+
+    for (int j = 0; j < dim; j++)
+      dims_vec[i][j] = tls.tensor_shape_span(i)[j];
   }
 }
 
 template<>
 void Resize<GPUBackend>::RunImpl(DeviceWorkspace *ws, const int idx) {
-    const auto &input = ws->Input<GPUBackend>(idx);
-
-    auto &output = ws->Output<GPUBackend>(outputs_per_idx_ * idx);
-
-    ResizeParamDescr resizeDescr(this, resizeParam_->data());
-    DataDependentSetupGPU(input, output, batch_size_, false,
-                            inputImages(), outputImages(), NULL, &resizeDescr);
-
-    // Run the kernel
-    cudaStream_t old_stream = nppGetStream();
-    nppSetStream(ws->stream());
-    BatchedResize(
-        (const uint8**)input_ptrs_.data(),
-        batch_size_, C_, sizes(input_t).data(),
-        output_ptrs_.data(), sizes(output_t).data(),
-        resizeParam_->data(), getInterpType());
-    nppSetStream(old_stream);
-
-    // Setup and output the resize attributes if necessary
-    if (save_attrs_) {
-      TensorList<CPUBackend> attr_output_cpu;
-      vector<Dims> resize_shape(input.ntensor());
-
-      for (size_t i = 0; i < input.ntensor(); ++i) {
-        resize_shape[i] = Dims{2};
-      }
-
-      attr_output_cpu.Resize(resize_shape);
-
-      for (size_t i = 0; i < input.ntensor(); ++i) {
-        int *t = attr_output_cpu.mutable_tensor<int>(i);
-        t[0] = sizes(input_t).data()[i].height;
-        t[1] = sizes(input_t).data()[i].width;
-      }
-      ws->Output<GPUBackend>(outputs_per_idx_ * idx + 1).Copy(attr_output_cpu, ws->stream());
+  using Kernel = kernels::ResampleGPU<uint8_t, uint8_t>;
+  const auto &input = ws->Input<GPUBackend>(idx);
+  auto &kdata = kernel_data_.front();
+
+  kdata.context.gpu.stream = ws->stream();
+  kdata.requirements = Kernel::GetRequirements(
+      kdata.context,
+      view<const uint8_t, 3>(input),
+      resample_params_);
+  kdata.scratch_alloc.Reserve(kdata.requirements.scratch_sizes);
```

Comment:
Can we avoid reallocating it for every run?
How expensive if reallocating it for every run?

Reply:
`scratch_alloc` is persistent. It never shrinks.

Reply:
It could. We should provide option to presize it as well.